JPA Map storage concurrent execution of liquibase updates

### Describe the bug

There are currently two synchronization problems: 

The following code will add the current model to the list of `validatedModels`. A second request would not wait here while the Liquibase upgrade is performed, and would instead overtake the first request and would fail due to missing tables. 

https://github.com/keycloak/keycloak/blob/f6a2b334d1e0652dcaba76cf8fccdc464a3171cd/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L377-L398

A second problem arises as the databasechangelog table is not locked with Quarkus. With undertow, as long as the legacy store is in the classpath, the [CustomLockService](https://github.com/keycloak/keycloak/blob/7c1d6eae431be5bcea976e67e099c8fe4391118d/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/lock/CustomLockService.java) will take care of this. 

To untangle the old JPA legacy store from the new JPA map store, this needs to be duplicated / adjusted to the needs to JPA map storage.


### Version

999-SNAPSHOT

### Expected behavior

When one thread migrates the database, all other threads should wait for that to complete. 

### Actual behavior

Threads might overtake each other, and even try to do parallel deployments.

### How to Reproduce?

Debugging with the Quarkus distribution.

### Anything else?

_No response_

Another problem that might be related as notes by @DGuhr 
 
> [...] after creating the initial admin user and logging into the admin console, I created a new realm "foobar" and clicked through the menu to see what changeset gets applied when (old / default admin UI).
> 
> When I clicked on "groups" the first time, the following exception occured:
> 
> ```
> 2022-07-15 14:06:04,965 WARN  [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-2) SQL Error: 0, SQLState: 42P01
> 2022-07-15 14:06:04,965 ERROR [org.hibernate.engine.jdbc.spi.SqlExceptionHelper] (executor-thread-2) ERROR: relation "kc_group" does not exist
>   Position: 215
> 2022-07-15 14:06:04,971 ERROR [org.keycloak.services.error.KeycloakErrorHandler] (executor-thread-2) Uncaught server error: org.hibernate.exception.SQLGrammarException: could not extract ResultSet
> 	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:103)
> 	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:37)
> 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
> 	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
> 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:67)
> 	at org.hibernate.loader.Loader.getResultSet(Loader.java:2322)
> 	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2075)
> 	at org.hibernate.loader.Loader.executeQueryStatement(Loader.java:2037)
> 	at org.hibernate.loader.Loader.scroll(Loader.java:2945)
> 	at org.hibernate.loader.hql.QueryLoader.scroll(QueryLoader.java:610)
> 	at org.hibernate.hql.internal.ast.QueryTranslatorImpl.scroll(QueryTranslatorImpl.java:452)
> 	at org.hibernate.engine.query.spi.HQLQueryPlan.performScroll(HQLQueryPlan.java:353)
> 	at org.hibernate.internal.SessionImpl.scroll(SessionImpl.java:1596)
> 	at org.hibernate.query.internal.AbstractProducedQuery.doScroll(AbstractProducedQuery.java:1588)
> 	at org.hibernate.query.internal.AbstractProducedQuery.scroll(AbstractProducedQuery.java:1574)
> 	at org.hibernate.query.internal.AbstractProducedQuery.stream(AbstractProducedQuery.java:1598)
> 	at org.hibernate.query.criteria.internal.compile.CriteriaQueryTypeQueryAdapter.stream(CriteriaQueryTypeQueryAdapter.java:91)
> 	at org.hibernate.query.Query.getResultStream(Query.java:1140)
> 	at org.keycloak.models.map.storage.jpa.JpaMapKeycloakTransaction.read(JpaMapKeycloakTransaction.java:167)
> 	at org.keycloak.models.map.group.MapGroupProvider.getGroupsStreamInternal(MapGroupProvider.java:104)
> 	at org.keycloak.models.map.group.MapGroupProvider.getTopLevelGroupsStream(MapGroupProvider.java:162)
> 	at org.keycloak.models.map.realm.MapRealmAdapter.getTopLevelGroupsStream(MapRealmAdapter.java:1424)
> 	at org.keycloak.models.utils.ModelToRepresentation.toGroupHierarchy(ModelToRepresentation.java:169)
> 	at org.keycloak.services.resources.admin.GroupsResource.getGroups(GroupsResource.java:86)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
> 	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
> 	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
> 	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
> 	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
> 	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
> 	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
> 	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
> 	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
> 	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
> 	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
> 	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
> 	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
> 	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
> 	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
> 	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
> 	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
> 	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
> 	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
> 	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
> 	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
> 	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
> 	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:67)
> 	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:55)
> 	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
> 	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
> 	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
> 	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:380)
> 	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$5.handle(VertxHttpRecorder.java:358)
> 	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1212)
> 	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:163)
> 	at io.vertx.ext.web.impl.RoutingContextImpl.next(RoutingContextImpl.java:141)
> 	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:71)
> 	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
> 	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
> 	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
> 	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$13.runWith(VertxCoreRecorder.java:545)
> 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
> 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
> 	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
> 	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
> 	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> Caused by: org.postgresql.util.PSQLException: ERROR: relation "kc_group" does not exist
>   Position: 215
> 	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
> 	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
> 	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
> 	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
> 	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
> 	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
> 	at org.postgresql.jdbc.PgPreparedStatement.executeQuery(PgPreparedStatement.java:118)
> 	at io.agroal.pool.wrapper.PreparedStatementWrapper.executeQuery(PreparedStatementWrapper.java:78)
> 	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.extract(ResultSetReturnImpl.java:57)
> 	... 72 more
> 
> 2022-07-15 14:06:04,987 INFO  [org.keycloak.models.map.storage.jpa.liquibase.updater.MapJpaLiquibaseUpdaterProvider] (executor-thread-4) Updating database. Using changelog META-INF/jpa-groups-changelog.xml
> ```
> 
> as you can see from the log, it seems the changeset gets applied _after_ keycloak tries to access the groups table, blindly guessing: these are run on different threads, not seeing each other.
> 
> [...] I was still able to access the groups page, and also able to create a new group, and switching back to other settings and then to group the exception was gone at the second time.

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/lock/LiquibaseDBLockProviderFactory.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/lock/LiquibaseDBLockProviderFactory.java
@@ -20,17 +20,19 @@ package org.keycloak.connections.jpa.updater.liquibase.lock;
 import java.util.List;
 import org.jboss.logging.Logger;
 import org.keycloak.Config;
+import org.keycloak.common.Profile;
 import org.keycloak.common.util.Time;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.KeycloakSessionFactory;
 import org.keycloak.models.dblock.DBLockProviderFactory;
+import org.keycloak.provider.EnvironmentDependentProviderFactory;
 import org.keycloak.provider.ProviderConfigProperty;
 import org.keycloak.provider.ProviderConfigurationBuilder;
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
  */
-public class LiquibaseDBLockProviderFactory implements DBLockProviderFactory {
+public class LiquibaseDBLockProviderFactory implements DBLockProviderFactory, EnvironmentDependentProviderFactory {
 
     private static final Logger logger = Logger.getLogger(LiquibaseDBLockProviderFactory.class);
     public static final int PROVIDER_PRIORITY = 1;
@@ -86,5 +88,10 @@ public class LiquibaseDBLockProviderFactory implements DBLockProviderFactory {
                 .type("int")
                 .helpText("The maximum time to wait when waiting to release a database lock.")
                 .add().build();
+    }
+
+    @Override
+    public boolean isSupported() {
+        return !Profile.isFeatureEnabled(Profile.Feature.MAP_STORAGE);
     }
 }

--- a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java
+++ b/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java
@@ -166,6 +166,10 @@ public class JpaMapStorageProviderFactory implements
     private final String sessionProviderKey;
     private final String sessionTxKey;
 
+    // Object instances for each single JpaMapStorageProviderFactory instance per model type.
+    // Used to synchronize on when validating the model type area.
+    private final ConcurrentHashMap<Class<?>, Object> SYNC_MODELS = new ConcurrentHashMap<>();
+
     public final static DeepCloner CLONER = new DeepCloner.Builder()
         //auth-sessions
         .constructor(JpaRootAuthenticationSessionEntity.class,  JpaRootAuthenticationSessionEntity::new)
@@ -374,24 +378,29 @@ public class JpaMapStorageProviderFactory implements
 
         if (this.validatedModelNames.add(modelName)) {
         */
-        if (this.validatedModels.add(modelType)) {
-            Connection connection = getConnection();
-            try {
-                if (logger.isDebugEnabled()) printOperationalInfo(connection);
-
-                MapJpaUpdaterProvider updater = session.getProvider(MapJpaUpdaterProvider.class);
-                MapJpaUpdaterProvider.Status status = updater.validate(modelType, connection, config.get("schema"));
-
-                if (!status.equals(VALID)) {
-                    update(modelType, connection, session);
-                }
-            } finally {
-                if (connection != null) {
+        if (!this.validatedModels.contains(modelType)) {
+            synchronized (SYNC_MODELS.computeIfAbsent(modelType, mc -> new Object())) {
+                if (!this.validatedModels.contains(modelType)) {
+                    Connection connection = getConnection();
                     try {
-                        connection.close();
-                    } catch (SQLException e) {
-                        logger.warn("Can't close connection", e);
+                        if (logger.isDebugEnabled()) printOperationalInfo(connection);
+
+                        MapJpaUpdaterProvider updater = session.getProvider(MapJpaUpdaterProvider.class);
+                        MapJpaUpdaterProvider.Status status = updater.validate(modelType, connection, config.get("schema"));
+
+                        if (!status.equals(VALID)) {
+                            update(modelType, connection, session);
+                        }
+                    } finally {
+                        if (connection != null) {
+                            try {
+                                connection.close();
+                            } catch (SQLException e) {
+                                logger.warn("Can't close connection", e);
+                            }
+                        }
                     }
+                    validatedModels.add(modelType);
                 }
             }
         }

--- a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/connection/DefaultLiquibaseConnectionProvider.java
+++ b/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/connection/DefaultLiquibaseConnectionProvider.java
@@ -19,15 +19,20 @@ package org.keycloak.models.map.storage.jpa.liquibase.connection;
 
 import java.sql.Connection;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 import liquibase.Liquibase;
 import liquibase.Scope;
 import liquibase.database.Database;
 import liquibase.database.DatabaseFactory;
 import liquibase.exception.LiquibaseException;
+import liquibase.exception.ServiceNotFoundException;
 import liquibase.resource.ClassLoaderResourceAccessor;
 import liquibase.resource.ResourceAccessor;
+import liquibase.servicelocator.ServiceLocator;
+import liquibase.sqlgenerator.SqlGeneratorFactory;
 import liquibase.ui.LoggerUIService;
 import org.jboss.logging.Logger;
 import org.keycloak.models.KeycloakSession;
@@ -46,6 +51,7 @@ import org.keycloak.models.KeycloakSession;
 public class DefaultLiquibaseConnectionProvider implements MapLiquibaseConnectionProvider {
 
     private static final Logger logger = Logger.getLogger(DefaultLiquibaseConnectionProvider.class);
+    private static final String KEYCLOAK_JPA_LEGACY_MODULE = "org.keycloak.connections.jpa";
 
     @SuppressWarnings("unused")
     public DefaultLiquibaseConnectionProvider(final KeycloakSession session) {
@@ -90,6 +96,10 @@ public class DefaultLiquibaseConnectionProvider implements MapLiquibaseConnectio
             Scope.exit(scopeId);
         } catch (Exception e) {
             throw new RuntimeException("Failed to exist scope: " + e.getMessage(), e);
+        } finally {
+            // To avoid cached instances from the just closed scope.
+            // Must be called only after exiting the scope.
+            SqlGeneratorFactory.reset();
         }
     }
 
@@ -98,10 +108,34 @@ public class DefaultLiquibaseConnectionProvider implements MapLiquibaseConnectio
         final Map<String, Object> scopeValues = new HashMap<>();
         // Setting the LoggerUIService here prevents Liquibase from logging each change set to the console using java.util.Logging in the Quarkus setup
         scopeValues.put(Scope.Attr.ui.name(), new LoggerUIService());
+
+        // This uses the same serviceloader as the parent (that is replaced in Keycloak Quarkus)
+        // and prevents any class from the legacy JPA store to leak into the service discovery.
+        // This can be removed once the legacy JPA provider is no longer around in both the
+        // Wildfly and Quarkus distributions of Keycloak.
+        ServiceLocator serviceLocator = Scope.getCurrentScope().getServiceLocator();
+        scopeValues.put(Scope.Attr.serviceLocator.name(), new ServiceLocator() {
+            @Override
+            public int getPriority() {
+                return serviceLocator.getPriority();
+            }
+
+            @Override
+            public <T> List<T> findInstances(Class<T> interfaceType) throws ServiceNotFoundException {
+                List<T> instances = serviceLocator.findInstances(interfaceType);
+                // prevent any class from the legacy JPA store to leak into the service discovery
+                instances = instances.stream().filter(t -> !t.getClass().getPackage().getName().startsWith(KEYCLOAK_JPA_LEGACY_MODULE)).collect(Collectors.toList());
+                return instances;
+            }
+        });
         try {
             scopeId = Scope.enter(scopeValues);
         } catch (Exception e) {
             throw new RuntimeException("Failed to initialize Liquibase: " + e.getMessage(), e);
+        } finally {
+            // To avoid cached instances from another scope.
+            // Must be called only after entering the scope.
+            SqlGeneratorFactory.reset();
         }
         return scopeId;
     }

--- a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java
+++ b/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/liquibase/updater/MapJpaLiquibaseUpdaterProvider.java
@@ -42,23 +42,6 @@ import org.keycloak.models.map.storage.jpa.updater.MapJpaUpdaterProvider;
 
 public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
 
-    public static class ThreadLocalSessionContext {
-
-        private static final ThreadLocal<KeycloakSession> currentSession = new ThreadLocal<KeycloakSession>();
-    
-        public static KeycloakSession getCurrentSession() {
-            return currentSession.get();
-        }
-    
-        public static void setCurrentSession(KeycloakSession session) {
-            currentSession.set(session);
-        }
-    
-        public static void removeCurrentSession() {
-            currentSession.remove();
-        }
-    }
-    
     private static final Logger logger = Logger.getLogger(MapJpaLiquibaseUpdaterProvider.class);
 
     private final KeycloakSession session;
@@ -69,6 +52,8 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
 
     @Override
     public void update(Class<?> modelType, Connection connection, String defaultSchema) {
+        // Liquibase has a global Scopes / a global ScopeManager by default, and SqlGeneratorFactory is always global
+        // therefore, ensure that only one migration runs at a time
         synchronized (MapJpaLiquibaseUpdaterProvider.class) {
             this.updateSynch(modelType, connection, null, defaultSchema);
         }
@@ -76,6 +61,8 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
 
     @Override
     public void export(Class<?> modelType, Connection connection, String defaultSchema, File file) {
+        // Liquibase has a global Scopes / a global ScopeManager by default, and SqlGeneratorFactory is always global
+        // therefore, ensure that only one migration runs at a time
         synchronized (MapJpaLiquibaseUpdaterProvider.class) {
             this.updateSynch(modelType, connection, file, defaultSchema);
         }
@@ -83,9 +70,6 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
 
     protected void updateSynch(Class<?> modelType, Connection connection, File file, String defaultSchema) {
         logger.debug("Starting database update");
-
-        // Need ThreadLocal as liquibase doesn't seem to have API to inject custom objects into tasks
-        ThreadLocalSessionContext.setCurrentSession(session);
 
         Writer exportWriter = null;
         try (Liquibase liquibase = getLiquibase(modelType, connection, defaultSchema)) {
@@ -100,7 +84,6 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
             logger.error("Error has occurred while updating the database", e);
             throw new RuntimeException("Failed to update database", e);
         } finally {
-            ThreadLocalSessionContext.removeCurrentSession();
             if (exportWriter != null) {
                 try {
                     exportWriter.close();
@@ -136,15 +119,16 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
     }
 
     @Override
-    public Status validate(Class modelType, Connection connection, String defaultSchema) {
+    public Status validate(Class<?> modelType, Connection connection, String defaultSchema) {
+        // Liquibase has a global Scopes / a global ScopeManager by default
+        // therefore, ensure that only one Scope of liquibase runs at a time
         synchronized (MapJpaLiquibaseUpdaterProvider.class) {
             return this.validateSynch(modelType, connection, defaultSchema);
         }
     }
 
-    protected Status validateSynch(final Class modelType, final Connection connection, final String defaultSchema) {
+    protected Status validateSynch(final Class<?> modelType, final Connection connection, final String defaultSchema) {
         logger.debug("Validating if database is updated");
-        ThreadLocalSessionContext.setCurrentSession(session);
 
         try (Liquibase liquibase = getLiquibase(modelType, connection, defaultSchema)) {
 
@@ -183,7 +167,7 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
         return liquibase.listUnrunChangeSets(null, new LabelExpression(), false);
     }
 
-    private Liquibase getLiquibase(Class modelType, Connection connection, String defaultSchema) throws LiquibaseException {
+    private Liquibase getLiquibase(Class<?> modelType, Connection connection, String defaultSchema) throws LiquibaseException {
         MapLiquibaseConnectionProvider liquibaseProvider = session.getProvider(MapLiquibaseConnectionProvider.class);
         String modelName = ModelEntityUtil.getModelName(modelType);
         if (modelName == null) {
@@ -209,10 +193,6 @@ public class MapJpaLiquibaseUpdaterProvider implements MapJpaUpdaterProvider {
 
     @Override
     public void close() {
-    }
-
-    public static String getTable(String table, String defaultSchema) {
-        return defaultSchema != null ? defaultSchema + "." + table : table;
     }
 
 }

--- a/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/LiquibaseProcessor.java
+++ b/quarkus/deployment/src/main/java/org/keycloak/quarkus/deployment/LiquibaseProcessor.java
@@ -3,6 +3,7 @@ package org.keycloak.quarkus.deployment;
 import java.lang.reflect.Modifier;
 import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.List;
@@ -11,10 +12,12 @@ import java.util.Set;
 
 import io.quarkus.agroal.spi.JdbcDataSourceBuildItem;
 import io.quarkus.deployment.builditem.CombinedIndexBuildItem;
+import liquibase.lockservice.StandardLockService;
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.ClassInfo;
 import org.jboss.jandex.DotName;
 import org.jboss.jandex.IndexView;
+import org.keycloak.config.StorageOptions;
 import org.keycloak.connections.jpa.updater.liquibase.lock.DummyLockService;
 
 import io.quarkus.deployment.annotations.BuildStep;
@@ -27,6 +30,10 @@ import liquibase.parser.core.xml.XMLChangeLogSAXParser;
 import liquibase.servicelocator.LiquibaseService;
 import liquibase.sqlgenerator.SqlGenerator;
 import org.keycloak.quarkus.runtime.KeycloakRecorder;
+
+import static org.keycloak.config.StorageOptions.STORAGE;
+import static org.keycloak.quarkus.runtime.configuration.Configuration.getOptionalValue;
+import static org.keycloak.quarkus.runtime.configuration.MicroProfileConfigProvider.NS_KEYCLOAK_PREFIX;
 
 class LiquibaseProcessor {
 
@@ -72,8 +79,12 @@ class LiquibaseProcessor {
             }
         }
 
-        services.put(LockService.class.getName(), Arrays.asList(DummyLockService.class.getName()));
-        services.put(ChangeLogParser.class.getName(), Arrays.asList(XMLChangeLogSAXParser.class.getName()));
+        if (StorageOptions.StorageType.jpa.name().equals(getOptionalValue(NS_KEYCLOAK_PREFIX.concat(STORAGE.getKey())).orElse(null))) {
+            services.put(LockService.class.getName(), Collections.singletonList(StandardLockService.class.getName()));
+        } else {
+            services.put(LockService.class.getName(), Collections.singletonList(DummyLockService.class.getName()));
+        }
+        services.put(ChangeLogParser.class.getName(), Collections.singletonList(XMLChangeLogSAXParser.class.getName()));
 
         recorder.configureLiquibase(services);
     }

--- a/testsuite/utils/pom.xml
+++ b/testsuite/utils/pom.xml
@@ -303,6 +303,7 @@
                         <configuration>
                             <systemProperties>
                                 <systemProperty><key>keycloak.profile.feature.map_storage</key><value>enabled</value></systemProperty>
+                                <systemProperty><key>keycloak.dblock.provider</key><value>none</value></systemProperty>
                                 <systemProperty><key>keycloak.realm.provider</key><value>map</value></systemProperty>
                                 <systemProperty><key>keycloak.client.provider</key><value>map</value></systemProperty>
                                 <systemProperty><key>keycloak.clientScope.provider</key><value>map</value></systemProperty>

